# Task 1: PySpark Exploration

In [0]:
from pyspark.sql import functions as F

trips = spark.read.table("hyf.nyc_yellow.raw_trips")
zones = spark.read.table("hyf.nyc_yellow.raw_zones")

## Question 1: Which pickup borough has the most trips?

In [0]:
borough_counts = (
    trips
    .join(
        zones,
        trips.pickup_location_id == zones.location_id,
        "inner"
    )
    .groupBy("borough")
    .agg(F.count("*").alias("trip_count"))
    .orderBy(F.col("trip_count").desc())
)

borough_counts.show(1, truncate=False)

+---------+----------+
|borough  |trip_count|
+---------+----------+
|Manhattan|112028489 |
+---------+----------+
only showing top 1 row



## Question 2: What is the average total amount per payment type?

In [0]:
payment_average = (
    trips
    .groupBy("payment_type")
    .agg(
        F.round(
            F.avg("total_amount"),
            2
        ).alias("avg_total_amount")
    )
    .orderBy("payment_type")
)

payment_average.show(truncate=False)

+------------+----------------+
|payment_type|avg_total_amount|
+------------+----------------+
|0           |23.49           |
|1           |30.0            |
|2           |23.75           |
|3           |9.03            |
|4           |2.16            |
|5           |14.89           |
+------------+----------------+



## Findings

The pickup borough with the most trips is **Manhattan**, with **112,028,489 trips**.

The average `total_amount` is **23.49** for payment type 0, **30.00** for type 1, **23.75** for type 2, **9.03** for type 3, **2.16** for type 4, and **14.89** for type 5.

## PySpark versus dbt SQL

I would use PySpark for large-scale exploration or transformations that require distributed processing, DataFrame operations, or custom Python logic. I would use dbt SQL for tested, documented, and version-controlled transformations that create reusable analytics models.